### ISSUE TITLE
Process being killed when training the FastSAM model

### ISSUE THREAD

**Issue by huangqiu15444** [ultralytics/ultralytics#21239]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Sorry to bother you. I want to know how much memory is required to train the FastSAM model. I've noticed that the larger the dataset, the higher the memory usage during training. Although I already have 128GB of memory, when I use more than about 230,000 data samples from the SA - 1B dataset for training, my memory is almost full.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @huangqiu15444, thank you for reaching out and for your interest in Ultralytics 🚀! We appreciate your detailed description.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

For questions about training custom models like FastSAM, please share as much relevant information as possible—such as your dataset structure, training command, logs, and environment details—to help us provide more targeted assistance. Also, make sure to review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

In the meantime, we recommend checking the [Docs](https://docs.ultralytics.com/) for information on memory management, dataset handling, and FastSAM-specific guidance.

Join the Ultralytics community for more support and discussion! For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth discussions, check out [Discourse](https://community.ultralytics.com/), or head over to the [Subreddit](https://www.reddit.com/r/Ultralytics/) to share your experiences.

## Upgrade

Please make sure you're running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will review your issue and assist you soon!

**Comment by Y-T-G** [comment]
The size of the dataset shouldn't affect memory usage unless you're caching to RAM.

**Comment by huangqiu15444** [comment]
> The size of the dataset shouldn't affect memory usage unless you're caching to RAM.

How to check and how to clear the RAM cache?

**Comment by huangqiu15444** [comment]
Here is my training code:

model.train(
            data="sa_debug.yaml", \
            epochs=60, \
            batch=16, \
            imgsz=640, \
            overlap_mask=False, \
            save=True, \
            save_period=5, \
            device='0',\
            project='fastsam', \
            name='train', 
            val=False,close_mosaic=10,workers=4,cache=False)

**Comment by Y-T-G** [comment]
Maybe a bug in the FastSAM code

**Comment by huangqiu15444** [comment]
> Maybe a bug in the FastSAM code

The training method is the same as that of YOLOv8-seg.


from ultralytics import YOLO
model = YOLO(model="yolov8x-seg.yaml", \
             )
model.train(data="sa.yaml", \
            epochs=100, \
            batch=32, \
            imgsz=1024, \
            overlap_mask=False, \
            save=True, \
            save_period=5, \
            device='0,1,2,3,4,5,6,7',\
            project='fastsam', \
            name='test', 
            val=False,)

**Comment by glenn-jocher** [comment]
Since you already have `cache=False` set, the memory scaling with dataset size suggests a potential issue with data loading or memory management during training. Try reducing `workers=1` to minimize multiprocessing overhead, and consider using a smaller `batch` size initially to see if that helps isolate whether it's a batch-related memory issue or a data loading problem with large datasets.

**Comment by huangqiu15444** [comment]
> Since you already have `cache=False` set, the memory scaling with dataset size suggests a potential issue with data loading or memory management during training. Try reducing `workers=1` to minimize multiprocessing overhead, and consider using a smaller `batch` size initially to see if that helps isolate whether it's a batch-related memory issue or a data loading problem with large datasets.

Thank you very much. Some of the methods you recommended have had some effects, but they haven't fundamentally solved my problem. Currently, I can only reduce a portion of the training data. The biggest issue regarding memory usage is actually the label cache. My data format is the same as that of COCO8-seg. The more data there is, the more memory is required to cache the labels.

**Comment by glenn-jocher** [comment]
You're correct that label caching is the main memory bottleneck with large segmentation datasets. The `labels.cache` file stores all segmentation masks and annotations in memory for faster access during training. For SA-1B scale datasets, this can easily consume 100+ GB of RAM.

You can try using the `fraction` parameter to train on a subset of your data, like `fraction=0.5` to use 50% of the dataset, which should proportionally reduce the label cache memory usage. Alternatively, you could delete the existing `.cache` file in your dataset directory to force regeneration, though this won't change the fundamental memory requirements for the full dataset.
